doc/user_mgmt/user_mgmt_hld.md

```diff
@@ -0,0 +1,595 @@
+![role_flow](role_flow.png)
+#####  5.6.2 <a name='FlowDescription::'></a>Flow description:
+Usercfgd will get the "primary_group" and "secondary_groups" of the new role from ROLE_TABLE and run:
+
+- usermod  --gid {primary_group} {user_name}
+- usermod  --groups {secondary_groups} {user_name}
+
+###  5.7 <a name='UserPassword'></a>User password
+#####  5.7.1 <a name='FlowDigram:'></a>Flow digram:
+![password_flow](password_flow.png)
+#####  5.7.2 <a name='FlowDescription::'></a>Flow description:
+User will configure a plain-text password using config command.<br/>
+We will check using passwordutil if it is aligned with password hardening poilicies ( if passwh is enabled).<br/>
+if so the password will be hashed and written to USER_TABLE.<br/>
+
+Usercfgd will apply the hashed password by:
+
+- Calling the linux command: "usermod --password {password} {user_name}"
+- Adding the hashed-password to list of user passwords: “hashed-password_history” in USER_TABLE
+- update date of last password change: “passwd_date_of_last_change” in USER_TABLE
+
+
+###  5.8 <a name='UserPasswordExpiration'></a>User password expiration
+Usercfgd will bypass password expiration data by keeping it in config db.<br/>
+passwd_date_of_last_change will be updated in config db after each password update and applied on system after db migration.<br/>
+This will preserve the password expiration date after upgrade.<br/>
+If passwd_date_of_last_change field does not exist in config db on init daemon will take it from the system to support password expiration as part of password hardening rules.
+
+###  5.9 <a name='UserPasswdrdPrompt'></a>User password prompt
+Password change via ssh connection will be supported by replacing passwd binary by python wrapper.<br/>
+Wrapper will handle password prompt during ssh connection to apply password using user management deamon.<br/>
+In this case user management will check password hardening rules before applying.
+The motivation is to keep password in config db after password change.
+
+![password_prompt_flow](password_prompt_flow.png)
+
+
+
+
+###  6 <a name='SAIAPI'></a>SAI API
+No change.
+
+
+
+###  7 <a name='WarmbootandFastbootDesignImpact'></a>Warmboot and Fastboot Design Impact
+Not relevant.
+
+###  8 <a name='RestrictionsLimitations'></a>Restrictions/Limitations
+
+##### Username/Role-name
+username/role-name max length is 32 and it begins with a letter or an underscore,
+followed by letters, digits, underscores, or dashes.<br/>
+They can end with a dollar sign.
+
+
+
+
+###  9 <a name='TestPlan'></a>Test Plan
+####  9.1 <a name='UnitTestcases'></a>Unit Test cases
+###### Configuration – good flow
+  - Verify default values
+  - Add/Delete user
+  - Enable/Disable user
+  - Configure role of user
+  - Configure password of user
+  - Check password last change date after configure password
+  - Configure full-name of user
+
+###### Configuration - Negative flow
+  - Change default users role
```

Comment:
I think the following tests would also be appropriate:
- Add user and attributes but do not perform config save.  Perform reboot and verify expected behavior (this would be applicable for attribute changes and user deletion)
- Add new user with new attributes, config save, then downgrade to older SONiC version and verify expected results

Reply:
I will add these tests to the list.